### PR TITLE
to_dict method handles empty list case

### LINKED ISSUE #411: Empty lists not indexable
The following line https://github.com/OpenIDC/pyoidc/blob/master/src/oic/oauth2/message.py#L322 will raise an indexing exception if a userinfo endpoint returns an attribute who's key, value is an empty list.  

Example:

nicknames: []

I initially found this error in flask_pyoidc:

Trace: 

```
  File "/Users/akrug/workspace/sso-dashboard/env/lib/python2.7/site-packages/flask_pyoidc/flask_pyoidc.py", line 110, in _handle_authentication_response
    print(userinfo)
  File "/Users/akrug/workspace/sso-dashboard/env/lib/python2.7/site-packages/oic/oauth2/message.py", line 631, in __str__
    return '{}'.format(self.to_dict())
  File "/Users/akrug/workspace/sso-dashboard/env/lib/python2.7/site-packages/oic/oauth2/message.py", line 318, in to_dict
    elif isinstance(val, list) and isinstance(val[0], Message):
IndexError: list index out of range
```

Here is a version of the profile that threw the exception:

```
{
  "family_name": "Krug",
  "given_name": "Andrew",
  "nickname": "Andrew Krug",
  "groups": [
    "mozilliansorg_iam-project",
    "mozilliansorg_cloud incident response",
    "mozilliansorg_cis_whitelist",
  ],
  "emails": [
    {
      "verified": true,
      "name": "mozillians-primary",
      "value": "akrug@mozilla.com",
      "primary": true
    }
  ],
  "email": "akrug@mozilla.com",
  "name": "Andrew Krug",
  "picture": "https://cdn.mozillians.org/media/uploads/sorl-cache/92/2e/922efe09c26b1a04c93b7580ffae866d.jpg",
  "_HRData": {
    "placeholder": "empty"
  },
  "user_metadata": {},
  "lastName": "Krug",
  "userName": "akrug",
  "firstName": "Andrew",
  "displayName": "Andrew J Krug",
  "active": true,
  "shirtSize": "Fitted Small",
  "preferredLanguage": "en_US",
  "timezone": "US/Pacific",
  "primaryEmail": "akrug@mozilla.com",
  "lastModified": "2017-08-18T16:48:58+00:00",
  "created": "2017-02-06T22:53:54+00:00",
  "uris": [
    {
      "verified": false,
      "name": "mozillians-GitHub-39660",
      "value": "https://github.com/andrewkrug",
      "primary": false
    }
  ],
  "nicknames": [],
  "email_verified": true,
  "clientID": "7wyIItkJX4t7vYEaDmGrwP9k2fBh5qWP",
  "updated_at": "2017-08-21T19:15:36.519Z",
  "user_id": "ad|Mozilla-LDAP|akrug",
  "identities": [
    {
      "user_id": "Mozilla-LDAP|akrug",
      "provider": "ad",
      "connection": "Mozilla-LDAP",
      "isSocial": false
    }
  ],
  "created_at": "2017-02-06T22:53:30.473Z",
  "multifactor": [
    "duo"
  ],
  "sub": "ad|Mozilla-LDAP|akrug",
  "https://sso.mozilla.com/claim/groups": [
    "mozilliansorg_iam-project",
    "mozilliansorg_cloud incident response",
    "mozilliansorg_cis_whitelist",
  ],
  "https://sso.mozilla.com/claim/emails": [
    {
      "verified": true,
      "name": "mozillians-primary",
      "value": "akrug@mozilla.com",
      "primary": true
    }
  ],
  "https://sso.mozilla.com/claim/_HRData": {
    "placeholder": "empty"
  }
}
```

Here is an attempt I've made to do two things:

1. Make the statement more resilient.  Return None instead of returning en exception.
2. Make the python more idomatic.

https://github.com/mozilla-iam/pyoidc/pull/1/files

Please let me know if you're interested in this PR and I'll open a pull to resolve this problem.

### PR DESCRIPTION
**Pull request by andrewkrug**
create test for success and failure conditions for oic profile containing empty lists

add changelog

Closes #411 

Thanks for contributing to this library! Please include the following check
list in your pull request submission (you can delete this message). If your
changes don't need a change log or documentation update, please ignore this.

- [x] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [x] The documentation has been updated, if necessary.

---

### CODE REVIEW

**Comment by andrewkrug** [comment]
Not sure why coveralls thinks coverage has fallen.  1 line of code change plus add one test for a method that was previously uncovered.  Can we merge this anyway?

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ The format is based on the [KeepAChangeLog] project.
 
 ### Fixed
 - [#405]: Fix generation of endpoint urls
+- [#411]: Empty lists not indexable
```
Comment (lwm) [inline CHANGELOG.md]: You're missing the link to #411 or?

File: tests/test_oauth2_message.py [hunk 2]
```diff
@@ -116,7 +116,8 @@ class DummyMessage(Message):
         "opt_int": SINGLE_OPTIONAL_INT,
         "opt_str_list": OPTIONAL_LIST_OF_STRINGS,
         "req_str_list": REQUIRED_LIST_OF_STRINGS,
-        "opt_json": SINGLE_OPTIONAL_JSON}
+        "opt_json": SINGLE_OPTIONAL_JSON
+    }
```
Comment (lwm) [inline tests/test_oauth2_message.py]: Opinionated unrelated style change. Risky ;) (Can leave, just joking).

File: tests/test_oauth2_message.py [hunk 3]
```diff
@@ -651,6 +652,17 @@ def test_to_jwe(keytype, alg, enc):
     assert msg1 == msg
 
 
+def test_to_dict():
+    msg = Message(a={'a': {'foo': {'bar': [{'bat': []}]}}})
```
Comment (lwm) [inline tests/test_oauth2_message.py]: I found this a hard to read at a glance, maybe something like:

``` python
contents =  {'a': {'a': {'foo': {'bar': [{'bat': []}]}}}}
msg = Message(contents)
assert msg.to_dict(lev=0) == contents

contents =  {'c': []}
msg = Message(contents)
assert msg.to_dict(lev=1) == contents

```

Comment (andrewkrug) [inline tests/test_oauth2_message.py]: Sure happy to accommodate.  #willfix as we like to say at Mozilla.

File: tests/test_oauth2_message.py (hunk 2 above)
Comment (andrewkrug) [inline tests/test_oauth2_message.py]: Couldn't resist ;)

File: CHANGELOG.md (hunk 1 above)
Comment (andrewkrug) [inline CHANGELOG.md]: Will link.

**Comment by andrewkrug** [comment]
@lwm all the tests should be in keeping with the style you've requested and I've added a link in the changelog to the aforementioned issue.

**Comment by lwm** [comment]
:cake:
